near-bindgen/src/collections/vector.rs

```diff
@@ -0,0 +1,413 @@
+const ERR_ELEMENT_DESERIALIZATION: &[u8] = b"Cannot deserialize element";
+const ERR_ELEMENT_SERIALIZATION: &[u8] = b"Cannot serialize element";
+const ERR_INDEX_OUT_OF_BOUNDS: &[u8] = b"Index out of bounds";
+
+/// An iterable implementation of vector that stores its content on the trie.
+/// Uses the following map: index -> element.
+#[derive(BorshSerialize, BorshDeserialize)]
+pub struct Vector<T> {
+    len: u64,
+    prefix: Vec<u8>,
+    #[borsh_skip]
+    el: PhantomData<T>,
+}
+
+impl<T> Vector<T> {
+    /// Returns the number of elements in the vector, also referred to as its size.
+    pub fn len(&self) -> u64 {
+        self.len
+    }
+
+    /// Returns `true` if the vector contains no elements.
+    pub fn is_empty(&self) -> bool {
+        self.len == 0
+    }
+
+    /// Create new vector with zero elements. Use `id` as a unique identifier on the trie.
+    pub fn new(id: Vec<u8>) -> Self {
+        Self { len: 0, prefix: id, el: PhantomData }
+    }
+
+    fn index_to_lookup_key(&self, index: u64) -> Vec<u8> {
+        let mut lookup_key = Vec::with_capacity(self.prefix.len() + size_of::<u64>());
+        lookup_key.extend_from_slice(&self.prefix);
+        lookup_key.extend_from_slice(&index.to_le_bytes());
+        lookup_key
+    }
+
+    /// Returns the serialized element by index or `None` if it is not present.
+    pub fn get_raw(&self, index: u64) -> Option<Vec<u8>> {
+        if index >= self.len {
+            return None;
+        }
+        let lookup_key = self.index_to_lookup_key(index);
+        match env::storage_read(&lookup_key) {
+            Some(raw_element) => Some(raw_element),
+            None => env::panic(ERR_INCONSISTENT_STATE),
+        }
+    }
+
+    /// Removes an element from the vector and returns it in serialized form.
+    /// The removed element is replaced by the last element of the vector.
+    /// Does not preserve ordering, but is `O(1)`.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `index` is out of bounds.
+    pub fn swap_remove_raw(&mut self, index: u64) -> Vec<u8> {
+        if index >= self.len {
+            env::panic(ERR_INDEX_OUT_OF_BOUNDS)
+        } else if index + 1 == self.len {
+            match self.pop_raw() {
+                Some(x) => x,
+                None => env::panic(ERR_INCONSISTENT_STATE),
+            }
+        } else {
+            let lookup_key = self.index_to_lookup_key(index);
+            let last_lookup_key = self.index_to_lookup_key(self.len - 1);
+            self.len -= 1;
+
+            // Swap the last element with the current one.
+            let raw_last_value = if env::storage_remove(&last_lookup_key) {
+                match env::storage_get_evicted() {
+                    Some(x) => x,
+                    None => env::panic(ERR_INCONSISTENT_STATE),
+                }
+            } else {
+                env::panic(ERR_INCONSISTENT_STATE)
```

Comment:
curious, how it compiles if the return type of `env::panic` is `()`?..

Reply:
The return type of `env::panic` is `!`